File: go/isolated.go
package headgate

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sort"
	"strings"
	"time"
)

const (
	IsolatedProtocolPrefix  = "HEADGATE/1 "
	defaultMaxProcessOutput = int64(64 * 1024)
)

// IsolatedRequest is the immutable versioned document written to the child stdin.
type IsolatedRequest struct {
	Version       uint32 `json:"version"`
	JobID         string `json:"job_id"`
	Kind          string `json:"kind"`
	SchemaVersion uint32 `json:"schema_version"`
	PayloadBase64 string `json:"payload_base64"`
	Queue         string `json:"queue"`
	PartitionKey  string `json:"partition_key"`
	RateClass     string `json:"rate_class"`
	Weight        uint32 `json:"weight"`
	Attempt       uint32 `json:"attempt"`
	CrashAttempt  uint32 `json:"crash_attempt"`
	MaxAttempts   uint32 `json:"max_attempts"`
	Fence         uint64 `json:"fence"`
	DeadlineMs    int64  `json:"deadline_ms"`
}

func (r IsolatedRequest) Payload() ([]byte, error) {
	return base64.StdEncoding.DecodeString(r.PayloadBase64)
}

type IsolatedOutcome string

const (
	IsolatedSuccess     IsolatedOutcome = "success"
	IsolatedRetry       IsolatedOutcome = "retry"
	IsolatedSkip        IsolatedOutcome = "skip"
	IsolatedRevoke      IsolatedOutcome = "revoke"
	IsolatedSnooze      IsolatedOutcome = "snooze"
	IsolatedRateLimited IsolatedOutcome = "rate_limited"
	IsolatedUndecodable IsolatedOutcome = "undecodable"
)

// IsolatedResponse follows IsolatedProtocolPrefix on one stdout line.
type IsolatedResponse struct {
	Version uint32          `json:"version"`
	Outcome IsolatedOutcome `json:"outcome"`
	Error   string          `json:"error,omitempty"`
	DelayMs int64           `json:"delay_ms,omitempty"`
}

// IsolatedProcessConfig is a fixed executable invocation. The parent environment is
// cleared by default so child handlers receive only explicitly configured values.
type IsolatedProcessConfig struct {
	Program        string
	Args           []string
	Env            map[string]string
	InheritEnv     bool
	MaxOutputBytes int64
}

// RegisterIsolated registers T's kind and aliases for child-process execution. Job
// bytes travel only through stdin; they are never interpolated into a shell command.
func RegisterIsolated[T Args](r *Registry, cfg IsolatedProcessConfig) error {
	if cfg.Program == "" {
		return errors.New("headgate: isolated process program must not be empty")
	}
	if cfg.MaxOutputBytes < 0 {
		return errors.New("headgate: isolated process max output must not be negative")
	}
	if cfg.MaxOutputBytes == 0 {
		cfg.MaxOutputBytes = defaultMaxProcessOutput
	}
	var zero T
	kinds := []string{zero.Kind()}
	if aliases, ok := any(zero).(Aliased); ok {
		kinds = append(kinds, aliases.KindAliases()...)
	}
	for _, kind := range kinds {
		if err := ValidateKind(kind); err != nil {
			return err
		}
		if _, exists := r.handlers[kind]; exists {
			return fmt.Errorf("headgate: kind %q is registered more than once", kind)
		}
	}
	handler := func(ctx context.Context, claim Claim) error {
		return executeIsolated(ctx, cfg, claim)
	}
	for _, kind := range kinds {
		r.handlers[kind] = handler
	}
	return nil
}

func executeIsolated(ctx context.Context, cfg IsolatedProcessConfig, claim Claim) error {
	if cfg.MaxOutputBytes == 0 {
		cfg.MaxOutputBytes = defaultMaxProcessOutput
	}
	e := claim.Envelope
	request := IsolatedRequest{
		Version: 1, JobID: e.ID, Kind: e.Kind, SchemaVersion: e.SchemaVersion,
		PayloadBase64: base64.StdEncoding.EncodeToString(e.Payload), Queue: e.Queue,
		PartitionKey: e.PartitionKey, RateClass: e.RateClass, Weight: EffectiveWeight(e.Weight),
		Attempt: e.Attempt, CrashAttempt: e.CrashAttempt, MaxAttempts: e.MaxAttempts,
		Fence: claim.Fence, DeadlineMs: e.DeadlineMs,
	}
	input, err := json.Marshal(request)
	if err != nil {
		return err
	}
	cmd := exec.CommandContext(ctx, cfg.Program, cfg.Args...)
	if cfg.InheritEnv {
		cmd.Env = os.Environ()
	} else {
		cmd.Env = []string{}
	}
	keys := make([]string, 0, len(cfg.Env))
	for key := range cfg.Env {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		cmd.Env = append(cmd.Env, key+"="+cfg.Env[key])
	}
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return err
	}
	if err := cmd.Start(); err != nil {
		return err
	}
	type readResult struct {
		bytes    []byte
		overflow bool
		err      error
	}
	outCh := make(chan readResult, 1)
	errCh := make(chan readResult, 1)
	go func() {
		b, overflow, err := readProcessOutput(stdout, cfg.MaxOutputBytes)
		outCh <- readResult{b, overflow, err}
	}()
	go func() {
		b, overflow, err := readProcessOutput(stderr, cfg.MaxOutputBytes)
		errCh <- readResult{b, overflow, err}
	}()
	writeCh := make(chan error, 1)
	go func() {
		_, err := io.Copy(stdin, bytes.NewReader(input))
		closeErr := stdin.Close()
		if err == nil {
			err = closeErr
		}
		writeCh <- err
	}()
	waitErr := cmd.Wait()
	writeErr := <-writeCh
	out := <-outCh
	errOut := <-errCh
	if ctx.Err() != nil {
		return ctx.Err()
	}
	if writeErr != nil && !errors.Is(writeErr, os.ErrClosed) {
		return writeErr
	}
	if out.err != nil {
		return out.err
	}
	if errOut.err != nil {
		return errOut.err
	}
	if out.overflow || errOut.overflow {
		return fmt.Errorf("headgate: isolated handler output exceeded %d bytes", cfg.MaxOutputBytes)
	}
	if waitErr != nil {
		return fmt.Errorf("headgate: isolated handler exited: %w: %s", waitErr, strings.TrimSpace(string(errOut.bytes)))
	}
	response, err := parseIsolatedResponse(out.bytes)
	if err != nil {
		return err
	}
	return isolatedResponseError(response)
}

func readProcessOutput(r io.Reader, max int64) ([]byte, bool, error) {
	var kept bytes.Buffer
	n, err := io.Copy(&kept, io.LimitReader(r, max+1))
	if err != nil {
		return nil, false, err
	}
	overflow := n > max
	if overflow {
		if _, err := io.Copy(io.Discard, r); err != nil {
			return nil, true, err
		}
		return kept.Bytes()[:max], true, nil
	}
	return kept.Bytes(), false, nil
}

func parseIsolatedResponse(stdout []byte) (IsolatedResponse, error) {
	lines := bytes.Split(stdout, []byte{'\n'})
	for i := len(lines) - 1; i >= 0; i-- {
		if !bytes.HasPrefix(lines[i], []byte(IsolatedProtocolPrefix)) {
			continue
		}
		var response IsolatedResponse
		if err := json.Unmarshal(lines[i][len(IsolatedProtocolPrefix):], &response); err != nil {
			return response, err
		}
		if response.Version != 1 {
			return response, fmt.Errorf("headgate: isolated handler response version %d is unsupported", response.Version)
		}
		return response, nil
	}
	return IsolatedResponse{}, errors.New("headgate: isolated handler emitted no HEADGATE/1 response")
}

func isolatedResponseError(response IsolatedResponse) error {
	switch response.Outcome {
	case IsolatedSuccess:
		return nil
	case IsolatedRetry:
		if response.Error == "" {
			response.Error = "isolated handler requested retry"
		}
		return errors.New(response.Error)
	case IsolatedSkip:
		return ErrSkipJob
	case IsolatedRevoke:
		return ErrRevokeJob
	case IsolatedRateLimited:
		return ErrRateLimited
	case IsolatedSnooze:
		if response.DelayMs > int64(^uint64(0)>>1)/int64(time.Millisecond) {
			return errors.New("headgate: isolated snooze delay is too large")
		}
		return Snooze(timeDurationMilliseconds(response.DelayMs))
	case IsolatedUndecodable:
		if response.Error == "" {
			response.Error = "isolated handler rejected payload"
		}
		return &UndecodableError{Cause: errors.New(response.Error)}
	default:
		return fmt.Errorf("headgate: isolated handler returned unknown outcome %q", response.Outcome)
	}
}

func timeDurationMilliseconds(ms int64) time.Duration {
	if ms <= 0 {
		return 0
	}
	return time.Duration(ms) * time.Millisecond
}

File: crates/headgate/src/isolated.rs
use std::{collections::BTreeMap, ffi::OsString, path::PathBuf, process::Stdio, time::Duration};

use base64::{Engine as _, engine::general_purpose::STANDARD};
use serde::{Deserialize, Serialize};
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWriteExt};

use crate::{BoxError, CodecError, Control, Envelope, ErasedHandler, HandlerFuture, JobCtx};

pub const ISOLATED_PROTOCOL_PREFIX: &str = "HEADGATE/1 ";
const DEFAULT_MAX_OUTPUT_BYTES: usize = 64 * 1024;

/// Immutable request written to an isolated handler's stdin as one JSON document.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct IsolatedRequest {
    pub version: u32,
    pub job_id: String,
    pub kind: String,
    pub schema_version: u32,
    pub payload_base64: String,
    pub queue: String,
    pub partition_key: String,
    pub rate_class: String,
    pub weight: u32,
    pub attempt: u32,
    pub crash_attempt: u32,
    pub max_attempts: u32,
    pub fence: u64,
    pub deadline_ms: i64,
}

impl IsolatedRequest {
    pub fn payload(&self) -> Result<Vec<u8>, base64::DecodeError> {
        STANDARD.decode(&self.payload_base64)
    }
}

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IsolatedOutcome {
    Success,
    Retry,
    Skip,
    Revoke,
    Snooze,
    RateLimited,
    Undecodable,
}

/// The child emits `HEADGATE/1 ` followed by this JSON object on one stdout line.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct IsolatedResponse {
    pub version: u32,
    pub outcome: IsolatedOutcome,
    #[serde(default)]
    pub error: String,
    #[serde(default)]
    pub delay_ms: i64,
}

/// A fixed child-process command. Environment inheritance is disabled by default so a
/// handler receives only explicitly configured values, not every worker secret.
#[derive(Clone, Debug)]
pub struct IsolatedProcess {
    pub program: PathBuf,
    pub args: Vec<OsString>,
    pub env: BTreeMap<OsString, OsString>,
    pub inherit_env: bool,
    pub max_output_bytes: usize,
}

impl IsolatedProcess {
    pub fn new(program: impl Into<PathBuf>) -> Self {
        Self {
            program: program.into(),
            args: Vec::new(),
            env: BTreeMap::new(),
            inherit_env: false,
            max_output_bytes: DEFAULT_MAX_OUTPUT_BYTES,
        }
    }

    pub fn arg(mut self, arg: impl Into<OsString>) -> Self {
        self.args.push(arg.into());
        self
    }

    pub fn env(mut self, key: impl Into<OsString>, value: impl Into<OsString>) -> Self {
        self.env.insert(key.into(), value.into());
        self
    }

    pub fn inherit_env(mut self, inherit: bool) -> Self {
        self.inherit_env = inherit;
        self
    }

    pub fn max_output_bytes(mut self, bytes: usize) -> Self {
        self.max_output_bytes = bytes;
        self
    }

    pub(crate) fn validate(&self) -> Result<(), String> {
        if self.program.as_os_str().is_empty() {
            return Err("isolated process program must not be empty".into());
        }
        if self.max_output_bytes == 0 {
            return Err("isolated process max_output_bytes must be greater than zero".into());
        }
        Ok(())
    }
}

pub(crate) struct IsolatedHandler {
    process: IsolatedProcess,
}

impl IsolatedHandler {
    pub(crate) fn new(process: IsolatedProcess) -> Self {
        Self { process }
    }
}

impl ErasedHandler for IsolatedHandler {
    fn call(&self, ctx: JobCtx, env: Envelope) -> HandlerFuture {
        let process = self.process.clone();
        Box::pin(async move { execute(process, ctx, env).await })
    }
}

async fn execute(process: IsolatedProcess, ctx: JobCtx, env: Envelope) -> Result<(), BoxError> {
    let request = IsolatedRequest {
        version: 1,
        job_id: env.id,
        kind: env.kind,
        schema_version: env.schema_version,
        payload_base64: STANDARD.encode(env.payload),
        queue: env.queue,
        partition_key: env.partition_key,
        rate_class: env.rate_class,
        weight: headgate_core::effective_weight(env.weight),
        attempt: env.attempt,
        crash_attempt: env.crash_attempt,
        max_attempts: env.max_attempts,
        fence: ctx.lease().fence,
        deadline_ms: env.deadline_ms,
    };
    execute_request(process, request).await
}

async fn execute_request(
    process: IsolatedProcess,
    request: IsolatedRequest,
) -> Result<(), BoxError> {
    let input = serde_json::to_vec(&request)?;

    let mut command = tokio::process::Command::new(&process.program);
    command
        .args(&process.args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true);
    if !process.inherit_env {
        command.env_clear();
    }
    command.envs(&process.env);
    let mut child = command.spawn()?;
    let mut stdin = child.stdin.take().expect("piped child stdin");
    let stdout = child.stdout.take().expect("piped child stdout");
    let stderr = child.stderr.take().expect("piped child stderr");
    let max = process.max_output_bytes;

    let write = async move {
        let result = stdin.write_all(&input).await;
        let _ = stdin.shutdown().await;
        result
    };
    let (write_result, stdout_result, stderr_result, status_result) = tokio::join!(
        write,
        read_bounded(stdout, max),
        read_bounded(stderr, max),
        child.wait(),
    );
    if let Err(error) = write_result {
        if error.kind() != std::io::ErrorKind::BrokenPipe {
            return Err(Box::new(error));
        }
    }
    let (stdout, stdout_overflow) = stdout_result?;
    let (stderr, stderr_overflow) = stderr_result?;
    if stdout_overflow || stderr_overflow {
        return Err(message_error(format!(
            "isolated handler output exceeded {max} bytes"
        )));
    }
    let status = status_result?;
    if !status.success() {
        let detail = String::from_utf8_lossy(&stderr);
        return Err(message_error(format!(
            "isolated handler exited with {status}: {}",
            detail.trim()
        )));
    }
    let response = parse_response(&stdout)?;
    response_result(response)
}

async fn read_bounded(
    mut reader: impl AsyncRead + Unpin,
    max: usize,
) -> std::io::Result<(Vec<u8>, bool)> {
    let mut bytes = Vec::with_capacity(max.min(8192));
    let mut limited = (&mut reader).take((max as u64).saturating_add(1));
    limited.read_to_end(&mut bytes).await?;
    let overflow = bytes.len() > max;
    if overflow {
        bytes.truncate(max);
        tokio::io::copy(&mut reader, &mut tokio::io::sink()).await?;
    }
    Ok((bytes, overflow))
}

fn parse_response(stdout: &[u8]) -> Result<IsolatedResponse, BoxError> {
    let prefix = ISOLATED_PROTOCOL_PREFIX.as_bytes();
    let line = stdout
        .split(|b| *b == b'\n')
        .rev()
        .find(|line| line.starts_with(prefix))
        .ok_or_else(|| message_error("isolated handler emitted no HEADGATE/1 response"))?;
    let response: IsolatedResponse = serde_json::from_slice(&line[prefix.len()..])?;
    if response.version != 1 {
        return Err(message_error(format!(
            "isolated handler response version {} is unsupported",
            response.version
        )));
    }
    Ok(response)
}

fn response_result(response: IsolatedResponse) -> Result<(), BoxError> {
    match response.outcome {
        IsolatedOutcome::Success => Ok(()),
        IsolatedOutcome::Retry => Err(message_error(if response.error.is_empty() {
            "isolated handler requested retry".into()
        } else {
            response.error
        })),
        IsolatedOutcome::Skip => Err(Box::new(Control::Skip)),
        IsolatedOutcome::Revoke => Err(Box::new(Control::Revoke)),
        IsolatedOutcome::RateLimited => Err(Box::new(Control::RateLimited)),
        IsolatedOutcome::Snooze => Err(Box::new(Control::Snooze(Duration::from_millis(
            u64::try_from(response.delay_ms).unwrap_or(0),
        )))),
        IsolatedOutcome::Undecodable => Err(Box::new(CodecError::Malformed(
            if response.error.is_empty() {
                "isolated handler rejected payload".into()
            } else {
                response.error
            },
        ))),
    }
}

fn message_error(message: impl Into<String>) -> BoxError {
    Box::new(std::io::Error::other(message.into()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{Read, Write};

    fn request(payload: &[u8]) -> IsolatedRequest {
        IsolatedRequest {
            version: 1,
            job_id: "isolated-1".into(),
            kind: "isolated:test".into(),
            schema_version: 1,
            payload_base64: STANDARD.encode(payload),
            queue: "default".into(),
            partition_key: "tenant".into(),
            rate_class: "api".into(),
            weight: 1,
            attempt: 2,
            crash_attempt: 1,
            max_attempts: 5,
            fence: 7,
            deadline_ms: 0,
        }
    }

    #[test]
    fn parses_prefixed_response_after_handler_logs() {
        let response = parse_response(
            b"ordinary child log\nHEADGATE/1 {\"version\":1,\"outcome\":\"success\"}\n",
        )
        .unwrap();
        assert_eq!(response.outcome, IsolatedOutcome::Success);
    }

    #[test]
    fn rejects_unknown_protocol_version() {
        let error =
            parse_response(b"HEADGATE/1 {\"version\":2,\"outcome\":\"success\"}\n").unwrap_err();
        assert!(error.to_string().contains("unsupported"));
    }

    #[test]
    fn maps_child_control_outcomes_to_runtime_controls() {
        let response = |outcome, delay_ms| IsolatedResponse {
            version: 1,
            outcome,
            error: String::new(),
            delay_ms,
        };
        assert!(response_result(response(IsolatedOutcome::Success, 0)).is_ok());
        assert!(
            response_result(response(IsolatedOutcome::Skip, 0))
                .unwrap_err()
                .downcast_ref::<Control>()
                .is_some_and(|c| *c == Control::Skip)
        );
        assert!(
            response_result(response(IsolatedOutcome::Snooze, 250))
                .unwrap_err()
                .downcast_ref::<Control>()
                .is_some_and(|c| *c == Control::Snooze(Duration::from_millis(250)))
        );
    }

    #[test]
    #[ignore]
    fn isolated_child_helper() {
        if std::env::var_os("HG_ISOLATED_HELPER").is_none() {
            return;
        }
        let mut input = Vec::new();
        std::io::stdin().read_to_end(&mut input).unwrap();
        let request: IsolatedRequest = serde_json::from_slice(&input).unwrap();
        assert_eq!(request.job_id, "isolated-1");
        assert_eq!(request.payload().unwrap(), b"ok");
        assert!(
            std::env::var_os("PATH").is_none(),
            "the isolated environment must be clear by default"
        );
        if request.kind == "isolated:sleep" {
            std::thread::sleep(Duration::from_secs(5));
        }
        writeln!(
            std::io::stdout(),
            "{ISOLATED_PROTOCOL_PREFIX}{{\"version\":1,\"outcome\":\"success\"}}"
        )
        .unwrap();
    }

    fn helper_process() -> IsolatedProcess {
        IsolatedProcess::new(std::env::current_exe().unwrap())
            .arg("--ignored")
            .arg("--exact")
            .arg("isolated::tests::isolated_child_helper")
            .arg("--nocapture")
            .env("HG_ISOLATED_HELPER", "1")
    }

    #[tokio::test]
    async fn executes_versioned_request_with_sanitized_environment() {
        execute_request(helper_process(), request(b"ok"))
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn dropping_attempt_future_kills_sleeping_child() {
        let mut sleepy = request(b"ok");
        sleepy.kind = "isolated:sleep".into();
        let timed = tokio::time::timeout(
            Duration::from_millis(100),
            execute_request(helper_process(), sleepy),
        )
        .await;
        assert!(
            timed.is_err(),
            "the helper must still be sleeping at the deadline"
        );
    }
}
